### ISSUE TITLE
Canonicalize DatasetRef, now with PeerID!

### ISSUE THREAD

**Issue by ramfox** [qri-io/qri#297]
Working on adding PeerID to the CanonoicalizeDatasetRef func:

https://github.com/qri-io/qri/blob/528920f0847c3435de92ff6279976f2f1f624c63/repo/ref.go#L222-L248

It feels like maybe the PeerID part belongs in CanonicalizePeername:
https://github.com/qri-io/qri/blob/528920f0847c3435de92ff6279976f2f1f624c63/repo/ref.go#L250-L261

new signature:
func CanonicalizePeer(r Repo, ref *DatasetRef) error
* if there is no PeerID or Peername given, just return nil
* if the dataset is local, it tries to flesh out the PeerID and Peername using the ID and Peername in Profile
* if it is not local, it tries to fleshout the PeerID and Peername using the ID and Peername in the repo.Peers() list
* if there is a mis-match, that is, PeerID and Peername are both given, but they don't match the Peername and PeerID in either Profile or the peers list, an error is returned

* make sure to add tests for our new CanonicalizePeer function!

**Comment by b5** [comment]
Ok, so I hate to be _that_ person, but I think our current solution needs a bit more resolving when it comes to string representations of DatasetRefs. Before going further I'd like to point out that this is hard because we're trying to solve [zookoo's triangle](https://en.wikipedia.org/wiki/Zooko%27s_triangle), which is hard. Human readable names are hard. Let's give ourselves permission to take time to experiment and get this right.

With the addition of the PeerID field, we need a way to include the PeerID when a DatasetRef is turned into a string. Currently we have this:
```
peername/datasetname@network/hash
```
If we call DatasetRef.String() with a ref that has a PeerID property, this info is lost. So we have to find somewhere to put PeerID in the string. To me this points to two concepts that need to be pulled apart, because one needs to go away and the other needs to be strengthened.

### 1.  Human names & hash combos as dataset identifiers for time and space
```
peername/datasetname    @    /network/hash 
^^^^^^^^^^^^^^^^^^^          ^^^^^^^^^^^^^^^^^^^^^^
dataset identifier           point in time and space
```

### 2. Human names as an _alias_ for a canonical Identifier
```
peername/datasetname    @    /network/hash 
^^^^^^^^^^^^^^^^^^^^         ^^^^^^^^^^^^
human-readable alias         identifier
```

The first is an _implicit outcome of poor design_ that we need to destroy with fire. Ok, concept 1 fits _indside_ 2 so we're not destroying it so much as fitting it properly. 

Whatever, the second is the one we want. So based on that, I think we should refactor `DatasetRef.String()` and `ParseDatasetRef(string) DatasetRef` to work with this string form:
```
peername/datasetname@peerId/network/hash
```

And we should add a new method to DatasetRef: `DatasetRef.Pretty() string` that spits out:
```
peername/datasetname
```

**Comment by ramfox** [comment]
cool!

For parsing the datasetRef, how do we want to handle ambiguous cases, if there is a PeerID but no Hash, or no hash but there is a PeerID?

This is clear to me:
```
@base58hash/string/base58hash_2
PeerID: base58hash
network: string
DatasetID: base58hash_2
```

This is clear to me:
```
@string/base58hash
network: string
DatasetID: base58hash
```

But this I'm not sure we can do anymore:
```
@base58hash
DatasetID: base58hash
network: ipfs
```

It seems like if there is a hash with no network, it should be the PeerID

**Comment by b5** [comment]
bingo. Hash with no network == PeerID. now to refer to a dataset you gotta `/ipfs/QmFoo`. We can use b58 detection & known network names to make proper assumptions.

This has the nice upside of peername / peerID's working kinda like... email?
```golang
DatasetRef{
  Peername: "lucille"
  PeerID: "QmFoo",
}.String() == "lucille@Qmfoo"
```

**Comment by b5** [comment]
closed by #298
